Lib/test/test_clinic.py

```diff
@@ -948,6 +948,72 @@ def test_parameters_required_after_star(self):
                 out = self.parse_function_should_fail(block)
                 self.assertIn(msg, out)
 
+    def test_depr_star_invalid_format_1(self):
+        out = self.parse_function_should_fail("""
+            module foo
+            foo.bar
+                this: int
+                * [from 3]
+            Docstring.
+        """)
+        expected = (
+            "Error on line 0:\n"
+            "Function 'bar': '* [from ...]' format expected to be "
+            "'<major.minor>', where 'major' and 'minor' are digits.\n"
+        )
+        self.assertEqual(out, expected)
+
+    def test_depr_star_invalid_format_2(self):
+        out = self.parse_function_should_fail("""
+            module foo
+            foo.bar
+                this: int
+                * [from a.b]
+            Docstring.
+        """)
+        expected = (
+            "Error on line 0:\n"
+            "Function 'bar', '* [from <major.minor>]': "
+            "'major' and 'minor' must be digits, not 'a' and 'b'\n"
+        )
+        self.assertEqual(out, expected)
+
+    def test_parameters_required_after_depr_star(self):
+        out = self.parse_function_should_fail("""
+            module foo
+            foo.bar
+                this: int
+                * [from 3.14]
+            Docstring.
+        """)
+        msg = "Function bar specifies '* [from ...]' without any parameters afterwards."
+        self.assertIn(msg, out)
+
+    def test_depr_star_must_come_before_star(self):
+        out = self.parse_function_should_fail("""
+            module foo
+            foo.bar
+                this: int
+                *
+                * [from 3.14]
+            Docstring.
+        """)
+        msg = "Function 'bar': '* [from ...]' must come before '*'"
+        self.assertIn(msg, out)
+
+    def test_duplicate_depr_star(self):
+        out = self.parse_function_should_fail("""
+            module foo
+            foo.bar
+                a: int
+                * [from 3.14]
+                * [from 3.14]
+                b: int
+            Docstring.
+        """)
+        msg = "Function 'bar' uses '[from ...]' more than once."
+        self.assertIn(msg, out)
+
     def test_single_slash(self):
```

Comment:
I think that we are missing the deprecation warning check here. Calling `test_deprecate_positional_use` with deprecated positional must produce a runtime deprecation warning. That's its main feature.

Reply:
It seems we could reuse some of the pegen test support code (see Tools/peg_generator/pegen, `compile_c_extension`, etc.)

Reply:
There are two scenarios we want to test:

- deprecation warning emitted in Python (we can do this now, using some tricks in _testclinic.c)
- C preprocessor warnings and error emitted at compile time (this needs a new framework, similar to what pegen does)


I suggest we start with the former. The latter would be nice to have, but it requires more invasive changes, so I suggest we do that in separate PRs post this PR.